ENH: Add `fast_matrix_market` to scipy.io

alugowski opened the pull request:
[`fast_matrix_market`](https://github.com/alugowski/fast_matrix_market) (FMM) is a C++ library with Python bindings to read/write sparse and dense matrices to MatrixMarket `.mtx` files. Basically a significantly faster version of `scipy.io.mmread` and `scipy.io.mmwrite`. I am the author.

It's currently on PyPI and Anaconda, with cibuildwheel wheels for Linux, macOS, and Windows.

I've gotten feedback about contributing it to SciPy, and I'd love for that to happen.

Folks on the mailing list suggested adding to `scipy.io._fast_matrix_market` and go from there. That's what this PR is.

To motivate the benefits of FMM, here is a plot comparing FMM and `_mmio` read speeds:
![parallel-scaling-python-read](https://github.com/scipy/scipy/assets/2730364/9236943f-e12d-46e5-90f2-358bc66380b5)

## Questions:

* How best to expose FMM to users? Currently a user would have to import from `scipy.io._fast_matrix_market`. Should this replace the scipy.io methods, be released alongside, or ??
* Are there any standards with how to specify authorship in the source files? Since I don't know any better I have kept the headers I use in my own repo, but happy to change to fit any SciPy standards.
* FMM's API is intentionally interchangeable with _mmio's. I have added a few things that I felt were missing from _mmio. These are listed below. Any thoughts on these are welcome.

### API extensions over scipy.io._mmio methods:
FMM supports a few things that the existing _mmio methods do not:

* both `mmread` and `mmwrite`: optional `parallelism` argument to specify thread pool size. Does this fit with how other scipy packages specify parallelism?
* `mmread`: optional `long_type` boolean. Enables loading values as `longdouble` and `longcomplex` instead of `float64` and `complex64`. AFAIK the _mmio methods cannot read data that requires `longdouble` precision, but can write it.
* `mmwrite`: the `symmetry` argument has a new default value of `AUTO`. `_mmio.mmwrite`'s default value is None, which looks at the matrix values for any symmetry. This is great, but is extremely slow on large matrices. It can easily slow down a write by 5x. The FMM solution is to still do that if the user explicitly asks for it (by passing `None`), but by default only do it on small matrices. This default can be overridden to match _mmio behavior by setting `fmm.ALWAYS_FIND_SYMMETRY = True`. All non-AUTO values behave just like in `_mmio.mmwrite`. All the benchmarks presented here use `symmetry='general'` to avoid this performance hit to `_mmio.mmwrite`.

Two improvements that are not visible in the API:
* `mmread`: automatically switch to 64-bit indices if the `.mtx` file's dimensions do not fit in 32-bit indices. `_mmio` uses `intc`, and on platforms where that is 32-bit `_mmio.mmread` simply crashes. Similarly on platforms where `intc` is 64-bit FMM will use 32-bit indices and save memory if the matrix dimensions allow it.
* `mmwrite`: if the matrix being written is a `csr_matrix` or `csc_matrix` then a dedicated implementation writes those matrices directly. `_mmio.mmwrite` converts to `coo_matrix` and writes that instead. This avoided copy is visible in the memory benchmark.

## Notes:

FMM fills in compiler support gaps with two libraries: [fast_float](https://github.com/fastfloat/fast_float) (used under MIT license), and [Ryu](https://github.com/ulfjack/ryu) (used under Boost license). See README.md for details. These compiler gaps are nearly filled in the latest compiler versions, but not the ones used to build wheels.

FMM passes the `test_mmio` test suite, and needs to continue doing so, so instead of adding another test suite I added a pytest fixture to `test_mmio` that runs all tests against both implementations. I added a test for `mmread`'s `long_type` argument.

I've also added an `io_mm` benchmark that runs for both `_mmio` and FMM. It is modeled on `io_matlab` and tests memory usage and runtime (on small 10MB matrices).

## Benchmark outputs:

For the `MemUsage` benchmark the unit is "actual/optimal memory usage ratio", which is what the io_matlab benchmark reports. This the ratio of peak process memory usage and matrix size in bytes.

Reading a CSR matrix makes no sense, so that column is 0.




```
· Discovering benchmarks
· Running 2 total benchmarks (1 commits * 1 environments * 2 benchmarks)
[  0.00%] ·· Benchmarking existing-py_opt_conda_envs_scipy-dev_bin_python
[ 50.00%] ··· io_mm.MemUsage.track_mmread                                                                                                                                                                                                   1/24 failed
[ 50.00%] ··· ====== ============================== ==================== ==================== =====
              --                                                      matrix_type
              ------------------------------------- -----------------------------------------------
               size          implementation                dense                 coo           csr
              ====== ============================== ==================== ==================== =====
                1M           scipy.io._mmio                92.968               91.932          0
                1M    scipy.io._fast_matrix_market        108.156              108.092          0
               10M           scipy.io._mmio               10.0648              10.0608          0
               10M    scipy.io._fast_matrix_market         15.95                15.454          0
               100M          scipy.io._mmio               1.89012              1.87936          0
               100M   scipy.io._fast_matrix_market         2.6274              2.63188          0
               300M          scipy.io._mmio          1.2808666666666666         failed          0
               300M   scipy.io._fast_matrix_market   1.5505066666666667   1.5493333333333332    0
              ====== ============================== ==================== ==================== =====

[ 50.00%] ···· For parameters: '300M', 'scipy.io._mmio', 'coo'


               asv: benchmark timed out (timeout 240s)

[100.00%] ··· io_mm.MemUsage.track_mmwrite                                                                                                                                                                                                  2/24 failed
[100.00%] ··· ====== ============================== ========= ==================== ====================
              --                                                        matrix_type
              ------------------------------------- ---------------------------------------------------
               size          implementation           dense           coo                  csr
              ====== ============================== ========= ==================== ====================
                1M           scipy.io._mmio           92.84          98.72                98.284
                1M    scipy.io._fast_matrix_market   105.084        107.512              107.684
               10M           scipy.io._mmio          10.1028        10.6968              12.2012
               10M    scipy.io._fast_matrix_market    11.356        11.3576              11.4472
               100M          scipy.io._mmio          1.88588        2.39664               3.8756
               100M   scipy.io._fast_matrix_market    2.0278        2.03196              2.04592
               300M          scipy.io._mmio           failed        1.78268               failed
               300M   scipy.io._fast_matrix_market   1.33392   1.3363733333333334   1.3624133333333333
              ====== ============================== ========= ==================== ====================

[100.00%] ···· For parameters: '300M', 'scipy.io._mmio', 'dense'


               asv: benchmark timed out (timeout 240s)

               For parameters: '300M', 'scipy.io._mmio', 'csr'


               asv: benchmark timed out (timeout 240s)

```

Another run for the IOSpeed benchmark:
```
[  0.00%] ·· Benchmarking existing-py_opt_homebrew_Caskroom_miniconda_base_envs_scipy-dev_bin_python
[ 12.50%] ··· Running (io_mm.IOSpeed.time_mmread--)..
[ 62.50%] ··· io_mm.IOSpeed.time_mmread                                                                                                                                                                                                 ok
[ 62.50%] ··· ============================== ============ ============ ============
              --                                          matrix_type
              ------------------------------ --------------------------------------
                      implementation            dense         coo          csr
              ============================== ============ ============ ============
                      scipy.io._mmio           544±2ms      539±2ms     86.6±0.7ns
               scipy.io._fast_matrix_market   48.3±0.8ms   40.9±0.6ms   86.7±0.7ns
              ============================== ============ ============ ============

[ 75.00%] ··· io_mm.IOSpeed.time_mmwrite                                                                                                                                                                                                ok
[ 75.00%] ··· ============================== ============ ========== ==========
              --                                        matrix_type
              ------------------------------ ----------------------------------
                      implementation            dense        coo        csr
              ============================== ============ ========== ==========
                      scipy.io._mmio           839±2ms     489±2ms    636±2ms
               scipy.io._fast_matrix_market   15.1±0.4ms   25.7±4ms   33.9±3ms
              ============================== ============ ========== ==========
```

alugowski commented:
Here is why the two dependencies are needed. FMM includes fallbacks to the standard libraries, but apart from being slower those routines also internally lock on the system locale. That kills parallelism.

Here is a `mmwrite` benchmark using standard library double-to-string conversion:

![parallel-scaling-python-write-stdlib](https://github.com/scipy/scipy/assets/2730364/a0e07279-0c8e-4dbb-b596-9e371d6c3bca)

Here is the same benchmark using Ryu:

![parallel-scaling-python-write-ryu](https://github.com/scipy/scipy/assets/2730364/038b8fe5-11ba-44ba-b43c-39358fd630b8)


Note that the Ryu version has a faster sequential speed (p=1), but more importantly it can be parallelized.

There is a similar story with `mmread` and fast_float.

alugowski commented:
That [circleci `build_scipy`](https://circleci.com/gh/scipy/scipy/69170) test fail run  has this message:

```
[1616/1624] Compiling C++ object scipy/optimize/_highs/_highs_wrapper.cpython-39-x86_64-linux-gnu.so.p/meson-generated__highs_wrapper.cpp.o
[1617/1624] Linking target scipy/optimize/_highs/_highs_wrapper.cpython-39-x86_64-linux-gnu.so
[1618/1624] Compiling C object scipy/io/matlab/_mio5_utils.cpython-39-x86_64-linux-gnu.so.p/meson-generated__mio5_utils.c.o
[1619/1624] Linking target scipy/io/matlab/_mio5_utils.cpython-39-x86_64-linux-gnu.so
[1620/1624] Compiling C object scipy/io/_fast_matrix_market/_core.cpython-39-x86_64-linux-gnu.so.p/fast_matrix_market_dependencies_ryu_ryu_f2s.c.o
[1621/1624] Compiling C object scipy/io/_fast_matrix_market/_core.cpython-39-x86_64-linux-gnu.so.p/fast_matrix_market_dependencies_ryu_ryu_d2s.c.o
[1622/1624] Compiling C object scipy/io/_fast_matrix_market/_core.cpython-39-x86_64-linux-gnu.so.p/fast_matrix_market_dependencies_ryu_ryu_d2fixed.c.o
ninja: build stopped: interrupted by user.

Too long with no output (exceeded 10m0s): context deadline exceeded
```

I'm not sure what to do with that information. I do see locally that the `d2fixed.c` compile is the very last line in `compile_commands.json`, so it's not clear to me whether that froze or whatever Ninja did next froze.

alugowski commented:
> ```
> ninja: build stopped: interrupted by user.
> 
> Too long with no output (exceeded 10m0s): context deadline exceeded
> ```

I believe this was caused by the core module being compiled in one step. I split it into multiple modules that can be compiled in parallel so hopefully that issue is resolved.

alugowski commented:
One more design choice: FMM supports two operations that `_mmio` does not:

* direct csr/csc writes, as mentioned in the OP
* reading `vector` Matrix Market files. These are non-standard and very rare, but exist in the wild. `_mmio.mmread` does not support these.

Together these two features account for about 25% of library size. One or both can dropped. I'll keep them in the stand-alone FMM package.

rgommers commented on scipy/io/_fast_matrix_market/meson.build:
```diff
@@ -0,0 +1,33 @@
+    'src/_core_write_coo.cpp',
+    'src/_core_write_csc.cpp',
+  ],
+  override_options: ['cpp_std=c++17'],
```
I'm getting consistent build errors locally on a macOS arm64 machine in a conda-forge environment with Clang 11.1.0:
```
../scipy/io/_fast_matrix_market/fast_matrix_market/include/fast_matrix_market/fast_matrix_market.hpp:29:32: error: constexpr variable 'kSpace' must be initialized by a constant expression
    constexpr std::string_view kSpace = " ";
```
My first guess is that it has something to do with this override. It may be problematic to force C++17 like this, see gh-18408 for remaining issues with defaulting to it.

alugowski replied:
Hmm. I just installed clang-11 from homebrew on my macOS arm64 machine and it compiles ok. I'll have to dig some more.

The problem in that issue is specific to `std::aligned_alloc` and I don't use that method. I'm also only enabling C++17 for this module and not globally, so everything else still builds as before. So far all the 64-bit tests pass on both macOS and Linux so the switch must be working. Hopefully no serious issues on Windows.

alugowski replied:
@rgommers The cause is that when building with conda's clang-11 there is an extra `-std=c++14` argument added *after* the `-std=c++17` that my meson line adds. Of course the last `-std` argument is used. This extra argument does not appear for the Homebrew clang11, or the default clang.

Maybe someone familiar with the build process can shed some light on why, and also if all conda clang versions are similarly affected.

Here are the build commands in `build/compile_commands.json` for one of the FMM files for the two Clang 11 versions:

Conda clang-11 build command

```
clang++ -Iscipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p -Iscipy/io/_fast_matrix_market -I../scipy/io/_fast_matrix_market -I../scipy/io/_fast_matrix_market/fast_matrix_market/include -I../../../../../../opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev-clang11/lib/python3.10/site-packages/numpy/core/include -I../scipy/io/_fast_matrix_market/fast_matrix_market/dependencies/fast_float/include -Iscipy/io/_fast_matrix_market/fast_matrix_market/dependencies/ryu -I../scipy/io/_fast_matrix_market/fast_matrix_market/dependencies/ryu -I/opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev-clang11/include -I/opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev-clang11/include/python3.10 -fvisibility=hidden -fvisibility-inlines-hidden -fcolor-diagnostics -Wall -Winvalid-pch -std=c++17 -O2 -g -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -std=c++14 -isystem /opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev-clang11/include -D_FORTIFY_SOURCE=2 -isystem /opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev-clang11/include -DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION -DFMM_FROM_CHARS_INT_SUPPORTED -DFMM_TO_CHARS_INT_SUPPORTED -DFMM_USE_FAST_FLOAT -DFMM_USE_RYU -MD -MQ scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core_write_csc.cpp.o -MF scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core_write_csc.cpp.o.d -o scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core_write_csc.cpp.o -c ../scipy/io/_fast_matrix_market/src/_core_write_csc.cpp
```


Homebrew clang-11 build command:

```
clang++ -Iscipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p -Iscipy/io/_fast_matrix_market -I../scipy/io/_fast_matrix_market -I../scipy/io/_fast_matrix_market/fast_matrix_market/include -I../../../../../../opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev/lib/python3.10/site-packages/numpy/core/include -I../scipy/io/_fast_matrix_market/fast_matrix_market/dependencies/fast_float/include -Iscipy/io/_fast_matrix_market/fast_matrix_market/dependencies/ryu -I../scipy/io/_fast_matrix_market/fast_matrix_market/dependencies/ryu -I/opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev/include -I/opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev/include/python3.10 -I/opt/homebrew/opt/llvm@11/include -fvisibility=hidden -fvisibility-inlines-hidden -fcolor-diagnostics -Wall -Winvalid-pch -std=c++17 -O2 -g -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem /opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev/include -DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION -DFMM_FROM_CHARS_INT_SUPPORTED -DFMM_TO_CHARS_INT_SUPPORTED -DFMM_USE_FAST_FLOAT -DFMM_USE_RYU -MD -MQ scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core_write_csc.cpp.o -MF scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core_write_csc.cpp.o.d -o scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core_write_csc.cpp.o -c ../scipy/io/_fast_matrix_market/src/_core_write_csc.cpp
```

alugowski replied:
This seems like it'll be a problem more broadly, so I opened a separate issue to track it: https://github.com/scipy/scipy/issues/18645

rgommers commented:
Thanks for the PR and detailed explanations on what this does and choices you made @alugowski! 

A small logistical comment: if it becomes annoying that GitHub Actions CI does not run without someone hitting the "Approve and run" button, you can make a small unrelated PR, some simple doc tweak or whatever, and we can merge that quickly to make the need for approving CI runs go away.

rgommers commented:
> How best to expose FMM to users? Currently a user would have to import from `scipy.io._fast_matrix_market`. Should this replace the scipy.io methods, be released alongside, or ??

FMM provides a superset of functionality and is basically much faster/better all around, right? If so, replacing the old implementation is perfectly fine, and preferred over keeping both.

> Are there any standards with how to specify authorship in the source files? Since I don't know any better I have kept the headers I use in my own repo, but happy to change to fit any SciPy standards.

We have no strict standards; SPDX headers are nice to have. I'm hoping we can standardize it at some point with REUSE (https://reuse.software/) tooling.

> FMM's API is intentionally interchangeable with `_mmio`'s. I have added a few things that I felt were missing from _mmio. These are listed below. Any thoughts on these are welcome.

From an initial reading of the API changes, the `parallelism` and `symmetry` ones sound good. The `long_type` one I'd like to look more at, since I'd like to get rid of `longdouble` in NumPy and SciPy sooner rather than later.

alugowski commented:
> A small logistical comment: if it becomes annoying that GitHub Actions CI does not run without someone hitting the "Approve and run" button, you can make a small unrelated PR, some simple doc tweak or whatever, and we can merge that quickly to make the need for approving CI runs go away.

That'd be nice. I just opened this one: https://github.com/scipy/scipy/pull/18640

rgommers commented:
Things build now for me. The build is unfortunately quite heavy: 

<img width="1201" alt="image" src="https://github.com/scipy/scipy/assets/98330/3106d746-ddb4-4459-8d88-2e23d283d611">

The three `ryu` files are in the 120-240ms range, but the 6 `_core` ones are all in the 9-20 second range. That seems too much, it'd be good to find ways to trim that down.

alugowski commented:
> Things build now for me. The build is unfortunately quite heavy:
> <img alt="image" width="1201" src="https://user-images.githubusercontent.com/98330/244269222-3106d746-ddb4-4459-8d88-2e23d283d611.png">
> 
> The three `ryu` files are in the 120-240ms range, but the 6 `_core` ones are all in the 9-20 second range. That seems too much, it'd be good to find ways to trim that down.

What tool is that from?

It's likely we're seeing the price of all the type combinations 74. There are so many to avoid casting, but there are a few long-hanging fruit to trim it down.
* 6 of those are reading `long double`, and 10 more for writing.
* Reading `vector` files I mentioned previously. These effectively count for about 15, mostly in the `_core_write_coo` chunk. Again `_mmio` doesn't support these, this would be new functionality.
* Writing `csr` accounts for 20. Again this is new functionality. Dropping this would remove the `_core_write_csc` chunk entirely.

If we went to just matching existing `_mmio` features then dropping all the above should cut the build drastically.

alugowski commented:
> > How best to expose FMM to users? Currently a user would have to import from `scipy.io._fast_matrix_market`. Should this replace the scipy.io methods, be released alongside, or ??
> 
> FMM provides a superset of functionality and is basically much faster/better all around, right? If so, replacing the old implementation is perfectly fine, and preferred over keeping both. 

Sounds good to me. The main downside I can think of is the library size, which shouldn't cause issues generally but might in some applications I'm not privvy to. The motivator for writing the library is large matrices, which are downright painful with `_mmio`. Small ones work fine with the relatively small Python code. There might also be some weird use case somewhere that the pure Python code can handle, like someone using a custom object for the `data` array or similar hacks.

> 
> > Are there any standards with how to specify authorship in the source files? Since I don't know any better I have kept the headers I use in my own repo, but happy to change to fit any SciPy standards.
> 
> We have no strict standards; SPDX headers are nice to have. I'm hoping we can standardize it at some point with REUSE (https://reuse.software/) tooling.
> 
Perfect. I already have SPDX headers in there. This makes things easier for me since I can do simple copy from my repo.

> > FMM's API is intentionally interchangeable with `_mmio`'s. I have added a few things that I felt were missing from _mmio. These are listed below. Any thoughts on these are welcome.
> 
> From an initial reading of the API changes, the `parallelism` and `symmetry` ones sound good. The `long_type` one I'd like to look more at, since I'd like to get rid of `longdouble` in NumPy and SciPy sooner rather than later.

If the type is going away then it doesn't make sense to try to get people to start using it. We could keep the ability to write `longdouble`.

Just curious, what makes you want to drop `longdouble`?

alugowski commented:
I'll need some help with the `setup.py`. From the logs it seems like the compiled extension is left in the wrong directory so the import fails. On my mac it appears to be correct.

rgommers commented:
> It's likely we're seeing the price of all the type combinations (74). There are so many to avoid casting, but there are a few long-hanging fruit to trim it down.

Let's try first to trim all of those indeed I'd say.



> Just curious, what makes you want to drop `longdouble`?

It's basically a useless alias to `double` on both Windows and macOS arm64, there's significant gaps in C99 support for it on more niche platforms (NetBSD and the like), and overall the ratio of its utility for a small set of use cases on 64-bit Linux is far from enough to justify the maintenance cost. See [here](https://mail.python.org/archives/list/numpy-discussion@python.org/message/LNSP7MFVDH53IDFB6AYEFEAXWO4MQQKG/) for more details.

alugowski commented:
I've implemented the changes above, and found a few more optimizations. The build time is significantly faster and the library size is also down to about a third of what it was.

![image](https://github.com/scipy/scipy/assets/2730364/893c5110-4b28-4c2f-8364-510972e89b03)

Curiously the `.so` that my scikit-build CMake workflow builds is half the size of the one meson builds, despite basically none of the pruning and an additional dependency.

rgommers commented:
> Curiously the `.so` that my scikit-build CMake workflow builds is half the size of the one meson builds, despite basically none of the pruning and an additional dependency.

Perhaps just a release vs a debug build? When using `dev.py` or `meson`/`ninja` directly, we use the `debugoptimized` build type. When going via `python -m build` or `pip`, we use a release build type. The difference is typically on the order of 3x-6x

alugowski commented:
> > Curiously the `.so` that my scikit-build CMake workflow builds is half the size of the one meson builds, despite basically none of the pruning and an additional dependency.
> 
> Perhaps just a release vs a debug build? When using `dev.py` or `meson`/`ninja` directly, we use the `debugoptimized` build type. When going via `python -m build` or `pip`, we use a release build type. The difference is typically on the order of 3x-6x

Good point, that sounds likely. I was comparing it to a release build.

alugowski commented:
> This is very close it looks like. The 32-bit build still seems unhappy:
> 
> ```
> /scipy/test/lib/python3.9/site-packages/pybind11/include/pybind11/detail/../detail/../pytypes.h:1575:69:   required from ‘pybind11::bytes::bytes(const char*, const SzType&) [with SzType = long long int; typename std::enable_if<std::is_integral<_Tp>::value, int>::type <anonymous> = 0]’
> ../scipy/io/_fast_matrix_market/src/pystreambuf.h:267:41:   required from here
> /scipy/test/lib/python3.9/site-packages/pybind11/include/pybind11/detail/../detail/common.h:484:35: error: static assertion failed: Implicit narrowing is not permitted.
>   484 |     static_assert(sizeof(IntType) <= sizeof(ssize_t), "Implicit narrowing is not permitted.");
> ```

Yes. I've a found a fix for that error, but I'm going through my own test suite first before pushing to this branch.

alugowski commented:
Any thoughts on that `refguide_check` failure? This is the output:
```
Running checks for 36 modules:
scipy.cluster .. scipy.cluster.hierarchy .................................................................. scipy.cluster.vq .......... scipy.constants .............................................................................................................................................................................. scipy.datasets .......Downloading file 'ascent.dat' from 'https://raw.githubusercontent.com/scipy/dataset-ascent/main/ascent.dat' to '/home/circleci/.cache/scipy-data'.
.Downloading file 'ecg.dat' from 'https://raw.githubusercontent.com/scipy/dataset-ecg/main/ecg.dat' to '/home/circleci/.cache/scipy-data'.
.Downloading file 'face.dat' from 'https://raw.githubusercontent.com/scipy/dataset-face/main/face.dat' to '/home/circleci/.cache/scipy-data'.
... scipy.fft .............................................................................. scipy.fftpack ............................................................. scipy.fftpack.convolve ......... scipy.integrate ................................................................ scipy.interpolate ...................................................................................................... scipy.io ......................terminate called after throwing an instance of 'pybind11::error_already_set'
  what():  OSError: Can't do nonzero cur-relative seeks
Aborted (core dumped)
Task Error - refguide-check => Command error: 'env 
PYTHONPATH=/home/circleci/repo/build-install/lib/python3.9/site-packages 
/home/circleci/.pyenv/versions/3.9.17/bin/python 
/home/circleci/repo/tools/refguide_check.py --doctests' returned 134

Exited with code exit status 2

```

I don't see that error when I run `python dev.py python tools/refguide_check.py` locally.

rgommers commented:
I can reproduce that locally on macOS with:
```bash
% python dev.py refguide-check -s io
scipy.io ......................libc++abi: terminating due to uncaught exception of type pybind11::error_already_set: OSError: Can't do nonzero cur-relative seeks
```
It looks like the problem is that in text (rather than binary) mode it's only allowed to seek from the beginning of the file in Python. From https://docs.python.org/3.11/tutorial/inputoutput.html#methods-of-file-objects: _"In text files (those opened without a b in the mode string), only seeks relative to the beginning of the file are allowed (the exception being seeking to the very file end with `seek(0, 2))` and the only valid offset values are those returned from the `f.tell()`, or zero. Any other offset value produces undefined behaviour."_

It looks like you are calling back into Python from C++ with `py_seek`, and that doesn't handle the above constraint. The `mmread` docstring contains an example using `StringIO`, which may highlight the problem.

alugowski commented:
> I can reproduce that locally on macOS with:
> 
> ```shell
> % python dev.py refguide-check -s io
> scipy.io ......................libc++abi: terminating due to uncaught exception of type pybind11::error_already_set: OSError: Can't do nonzero cur-relative seeks
> ```

Thanks! yes I can reproduce it now.
> 
> It looks like the problem is that in text (rather than binary) mode it's only allowed to seek from the beginning of the file in Python. From https://docs.python.org/3.11/tutorial/inputoutput.html#methods-of-file-objects: _"In text files (those opened without a b in the mode string), only seeks relative to the beginning of the file are allowed (the exception being seeking to the very file end with `seek(0, 2))` and the only valid offset values are those returned from the `f.tell()`, or zero. Any other offset value produces undefined behaviour."_
> 
> It looks like you are calling back into Python from C++ with `py_seek`, and that doesn't handle the above constraint. The `mmread` docstring contains an example using `StringIO`, which may highlight the problem.

Good lead, it's the `mminfo` docstring. Looks like `pystreambuf`, the class that adapts Python stream objects to C++ stream objects calls `py_seek` in its `sync` method. I'll have to investigate why. I do use StringIO extensively in tests and explicitly avoid seeks to enable 1-pass reads, so this one is a surprise.

alugowski commented:
Ok so the issue only occurred on `mminfo` because that method only reads the header then exits. All other methods read the entire stream. `test_mmio` uses temp files, which FMM opens in C++ and again sidesteps this error.

`pystreambuf` would seek on sync because its own buffering could make its and the backing Python stream's positions not the same.  Luckily I'm already forced to use a string-to-bytes adapter class (because the C++ code only works on bytes). So I overrode the `seek` method in this adapter to ignore these seeks and the issue is fixed.

alugowski commented:
I believe this is ready.

The only things I can think of to add:
 * Add a way to return sparse arrays instead of `coo_matrix`? This should probably be a separate issue, I could use feedback on what that method should look like.
 * Modify the benchmark? I tested the `_mmio` and `_fmm` modules explicitly to compare the two implementations as some folks on the mailing list expressed interest in memory usage. Maybe modify to either only test `scipy.io.mmread/write`?

I added an outline of what each C++ file does to the README to make the reviewer's life easier.

alugowski commented:
I updated the `io_mm` benchmark to include the user-facing `scipy.io` methods. That should be the only interesting one long-term anyway.

The two implementations, `scipy.io._mmio` and `scipy.io._fast_matrix_market`, were included to help reviewers of this PR. Obviously the `scipy.io` performance matches `_fast_matrix_market`'s.

dschult commented:
To return a `coo_array` instead of a `coo_matrix` just use `coo_array(..)` to build the array instead of `coo_matrix(...)`.  You aren't doing matrix multiply or other algebra -- just writing it to a file, so the interface is the same.

I might be missing some algebra somewhere, but it seems like a gloabl replace of coo_matrix with coo_array would do it (and similar for csr is desired).

alugowski commented:
> To return a `coo_array` instead of a `coo_matrix` just use `coo_array(..)` to build the array instead of `coo_matrix(...)`. You aren't doing matrix multiply or other algebra -- just writing it to a file, so the interface is the same.
> 
> I might be missing some algebra somewhere, but it seems like a gloabl replace of coo_matrix with coo_array would do it (and similar for csr is desired).

Right. The tricky bit is the signature of `mmread()`. Today `scipy.io.mmread()` returns a `coo_matrix` (nevermind the dense case). I doubt it's safe to change the return type.

dschult commented:
Hmmm... yes I see that its tricky... 
It'd be great to have the `fast` version return `coo_array` since it is new. 
The `coo_matrix` is deprecated even if it isn't removed yet.
But I understand that there may be other issues involved.

alugowski commented:
> Hmmm... yes I see that its tricky... It'd be great to have the `fast` version return `coo_array` since it is new. The `coo_matrix` is deprecated even if it isn't removed yet. But I understand that there may be other issues involved.

It's a new implementation of an old method. Per discussion above, FMM replaces the existing implementation from `_mmio.py`. See change to [scipy/io/__init__.py](https://github.com/scipy/scipy/pull/18631/files#diff-f0255849ed169d498a8e51370ed2cc948ccf960a35be954a4b44fc3dfccfe127):
```python
from ._mmio import mminfo, mmread, mmwrite
```
Replaced with
```python
from ._fast_matrix_market import mminfo, mmread, mmwrite
```

This way user code will simply get faster. 

I agree, if it was a new method it should just return `coo_array`. 

My thought was to implement whatever migration plan is for other similar methods (generators, other loaders) while we're touching this code. I just don't know what that plan is.

alugowski commented:
That linter crash is puzzling.

rgommers commented:
> It's a new implementation of an old method [...]

This is right, there should not be a change to the return type here, that is orthogonal to implementing a faster algorithm and should be done as part of the overall work on changing `*_matrix` to `*_array`.

dschult commented:
Yes -- since this is replacing code for an existing api we should leave the return value as it is. 
The matrix -> array shift of existing api is a bigger change and orthogonal to this.
Thanks. :}

alugowski commented:
I found the reason for the peak RSS difference between FMM and `_mmio` as shown in the read MemoryUsage benchmark. The file chunks, about 1MB in size, are stored in `std::string`s between read and parse. These strings are freed appropriately, but it seems the allocators don't always reuse the freed memory and instead let the RSS grow. This isn't necessarily bad as any memory pressure would trigger reuse, but it makes it hard for simple tooling to confirm good behavior. I added an object pool to reuse the chunks instead and now peak RSS reflects actual use.

Updated benchmark results follow.
Again the number is (peak memory usage (RSS) / matrix size in bytes), so the small 1MB matrix runs are dominated by Python memory usage while the larger ones are dominated by the matrix itself.

FMM memory usage is identical to `_mmio`'s, except FMM is significantly better at writing `coo` matrices.

```
[ 87.50%] ··· io_mm.MemUsage.track_mmread                                                                                                                                                                                                       ok
[ 87.50%] ··· ======= ================== ==================== ======================== ====================== ====================================== ====================================
              --                                                                          implementation / matrix_type
              ------- -------------------------------------------------------------------------------------------------------------------------------------------------------------------
                size   scipy.io / dense     scipy.io / coo     scipy.io._mmio / dense   scipy.io._mmio / coo   scipy.io._fast_matrix_market / dense   scipy.io._fast_matrix_market / coo
              ======= ================== ==================== ======================== ====================== ====================================== ====================================
                 1M       87.351296           88.690688              87.171072               87.699456                       87.53152                              87.98208
                10M       9.6894976           9.7349632              9.6534528               9.6776192                      9.6534528                             9.8250752
                100M      1.8761728           1.88370944             1.86896384              1.89288448                     1.86896384                            1.8778112
                300M      1.29265664      1.2984456533333333     1.2920558933333333          1.29953792                 1.2932573866666666                        1.29953792
                500M     1.179557888         1.181679616            1.177755648             1.182334976                    1.179557888                           1.183645696
               1000M     1.090502656         1.090842624            1.089961984             1.092399104                    1.089601536                           1.091497984
              ======= ================== ==================== ======================== ====================== ====================================== ====================================

[100.00%] ··· io_mm.MemUsage.track_mmwrite                                                                                                                                                                                                      ok
[100.00%] ··· ======= ==================== ================ ======================== ====================== ====================================== ====================================
              --                                                                         implementation / matrix_type
              ------- -----------------------------------------------------------------------------------------------------------------------------------------------------------------
                size    scipy.io / dense    scipy.io / coo   scipy.io._mmio / dense   scipy.io._mmio / coo   scipy.io._fast_matrix_market / dense   scipy.io._fast_matrix_market / coo
              ======= ==================== ================ ======================== ====================== ====================================== ====================================
                 1M        78.598144          78.598144            78.598144               78.598144                      78.598144                             78.598144
                10M        7.8598144          7.8598144            7.8598144               7.8598144                      7.8598144                             7.8598144
                100M       1.51691264         1.53051136           1.47144704              1.97558272                     1.5161344                             1.53288704
                300M   1.1740228266666666     1.17932032       1.1577480533333333          1.65781504                 1.1734357333333334                    1.1812317866666666
                500M      1.104166912        1.107197952          1.094475776             1.594998784                    1.104175104                           1.107484672
               1000M      1.052360704        1.054539776          1.047289856             1.547300864                    1.052176384                            1.0544128
              ======= ==================== ================ ======================== ====================== ====================================== ====================================
```

alugowski commented:
The two test fails are unrelated. One is a fickle arpack test and the other is a numpy nightly bug (see gh-18721)

alugowski commented:
Rebased to pull in the patch for https://github.com/scipy/scipy/issues/18736

alugowski commented:
CI fail unrelated, will be fixed when https://github.com/jupyter/nbformat/pull/369 is released.

alugowski commented:
@rgommers @larsoner

rgommers commented:
With this PR (rebased on `main` to ensure there are no other differences):
```
$ python -m build
$ ls -lh dist/
total 123M
-rw-r--r-- 1 rgommers rgommers 70M 12 jul 17:21 scipy-1.12.0.dev0-cp310-cp310-linux_x86_64.whl
-rw-r--r-- 1 rgommers rgommers 54M 12 jul 17:19 scipy-1.12.0.dev0.tar.gz
```

On `main`:
```
$ python -m build
$ ls -lh dist/
$ ls -lh dist/
total 119M
-rw-r--r-- 1 rgommers rgommers 66M 12 jul 17:29 scipy-1.12.0.dev0-cp310-cp310-linux_x86_64.whl
-rw-r--r-- 1 rgommers rgommers 54M 12 jul 17:27 scipy-1.12.0.dev0.tar.gz
```

My wheels are larger than the ones `cibuildwheel` produces (I can't remember if that's expected). More importantly, I'm not sure why `faster_matrix_market/_core.so` is so large for my release build, compared to the standalone package. Will try to figure that out.

rgommers commented:
I pushed a commit with `[wheel build]` in the first line of the commit message, that triggers all our wheel build machinery - which should give a more definitive answer about artifact sizes etc.

rgommers commented:
Good I triggered the wheel builds it looks like, the macOS x86-64 ones are failing:
```
    FAILED: scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core.cpp.o
    c++ -Iscipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p -Iscipy/io/_fast_matrix_market -I../../scipy/io/_fast_matrix_market -I../../scipy/io/_fast_matrix_market/fast_matrix_market/include -I../../../../../../../private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-build-env-yj77gq5q/overlay/lib/python3.10/site-packages/numpy/core/include -I../../scipy/io/_fast_matrix_market/fast_matrix_market/dependencies/fast_float/include -Iscipy/io/_fast_matrix_market/fast_matrix_market/dependencies/ryu -I../../scipy/io/_fast_matrix_market/fast_matrix_market/dependencies/ryu -I/Library/Frameworks/Python.framework/Versions/3.10/include/python3.10 -I/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-build-env-yj77gq5q/overlay/lib/python3.10/site-packages/pybind11/include -fvisibility=hidden -fvisibility-inlines-hidden -fcolor-diagnostics -DNDEBUG -Wall -Winvalid-pch -std=c++17 -O3 -DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION -DFMM_SCIPY_PRUNE -DFMM_FROM_CHARS_INT_SUPPORTED -DFMM_TO_CHARS_INT_SUPPORTED -DFMM_USE_FAST_FLOAT -DFMM_USE_RYU -MD -MQ scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core.cpp.o -MF scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core.cpp.o.d -o scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core.cpp.o -c ../../scipy/io/_fast_matrix_market/src/_core.cpp
    In file included from ../../scipy/io/_fast_matrix_market/src/_core.cpp:16:
    In file included from ../../scipy/io/_fast_matrix_market/fast_matrix_market/include/fast_matrix_market/app/../fast_matrix_market.hpp:210:
    ../../scipy/io/_fast_matrix_market/fast_matrix_market/include/fast_matrix_market/field_conv.hpp:366:44: error: 'to_chars<int, 0>' is unavailable: introduced in macOS 10.15
            std::to_chars_result result = std::to_chars(ret.data(), ret.data() + ret.size(), value);
                                               ^
    ../../scipy/io/_fast_matrix_market/fast_matrix_market/include/fast_matrix_market/field_conv.hpp:393:16: note: in instantiation of function template specialization 'fast_matrix_market::int_to_string<int>' requested here
            return int_to_string(value);
                   ^
    /Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/charconv:392:1: note: 'to_chars<int, 0>' has been explicitly marked unavailable here
    to_chars(char* __first, char* __last, _Tp __value)
    ^
    In file included from ../../scipy/io/_fast_matrix_market/src/_core.cpp:16:
    In file included from ../../scipy/io/_fast_matrix_market/fast_matrix_market/include/fast_matrix_market/app/../fast_matrix_market.hpp:210:
    ../../scipy/io/_fast_matrix_market/fast_matrix_market/include/fast_matrix_market/field_conv.hpp:366:44: error: 'to_chars<long long, 0>' is unavailable: introduced in macOS 10.15
            std::to_chars_result result = std::to_chars(ret.data(), ret.data() + ret.size(), value);
                                               ^
    ../../scipy/io/_fast_matrix_market/fast_matrix_market/include/fast_matrix_market/field_conv.hpp:397:16: note: in instantiation of function template specialization 'fast_matrix_market::int_to_string<long long>' requested here
            return int_to_string(value);
                   ^
    /Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/charconv:392:1: note: 'to_chars<long long, 0>' has been explicitly marked unavailable here
    to_chars(char* __first, char* __last, _Tp __value)
    ^
    2 errors generated.
```
Could you have a look at that @alugowski?

alugowski commented:
The macOS wheels appear to fail because the x86 wheels target macOS 10.9 and it looks like `<charconv>` support is added in 10.15. I pushed a commit to use `strtoll()` and friends on macOS instead. They appear to perform well so it shouldn't be an issue.

I would have preferred to do this with a meson `compiles()` check like I do in CMake so it would automatically switch the feature back on in the future when the target OS newer, but it seems Meson's `compiles()` won't let me set the c++ standard.

alugowski commented:
My cibuildwheel is bone standard, so I'm not sure why the library is so much bigger here despite being pruned quite a lot.

One annoying bit is that meson puts FMM at the very end of the build so basically no other compiles overlap. I could try to break up the files a bit more so it parallelizes better.

We could prune it further if you're uncomfortable with the size, but with a type conversion penalty at runtime. For example we could drop the 32-bit index methods, which would mean that users with such matrices would see an upcast to 64-bit ints at runtime. It would double memory requirements, but we could drop the core_write_coo_32 file altogether. Similary I see that `_mmio` does a type cast for the values array, visible as its 50% higher memory use in the coo write memory benchmark. If we did that it would again mean we could prune some more methods at the cost of a cast at runtime.

rgommers commented:
> One annoying bit is that meson puts FMM at the very end of the build so basically no other compiles overlap. I could try to break up the files a bit more so it parallelizes better.

This is determined by the order of `subdir`'s at the bottom of `scipy/meson.build`. `io` is right at the bottom because until now it was a very lightweight module. You can move it up, that should help - between `fft` and `spatial` is probably about right given the compile times involved.

rgommers commented:
> My cibuildwheel is bone standard, so I'm not sure why the library is so much bigger here despite being pruned quite a lot.

I think it's some difference in how I built the wheels. I checked the sizes reported by GHA, and it's similar to your standalone wheels. This PR ([log](https://github.com/scipy/scipy/actions/runs/5534254243#artifacts)):

<img width="463" alt="image" src="https://github.com/scipy/scipy/assets/98330/902b5c6c-d7ee-42f8-8981-a7d07eeff276">

Last scheduled wheel builder run on `main` [https://github.com/scipy/scipy/actions/runs/5493711416](log):

<img width="463" alt="image" src="https://github.com/scipy/scipy/assets/98330/eec506f4-2046-4b7d-9770-3c279517c1df">

That's ~700 kb differences - let's call that good, no need for trimming it further.

rgommers commented:
The last commits pulled in an unintended update of the `unuran` git submodule, could you revert that? You may want to trigger `[wheel build]` again, the GHA jobs didn't run because you pushed a "merge with main" commit right after which cancelled them.

alugowski commented:
Apologies for the git mess up, `git revert` of the extra merge commit did not do what I expected it to do. Working to resolve. It pulled in a lot of folks as reviewers, and I don't appear to be able to fix that, though.

alugowski commented:
Ok, I fixed my extraneous merge snafu.

Again apologies for all the folks accidentally added as reviewers. Looks like I don't have the ability to fix that, can someone with appropriate permissions remove all but the first two?

> This is determined by the order of `subdir`'s at the bottom of `scipy/meson.build`. `io` is right at the bottom because until now it was a very lightweight module. You can move it up, that should help - between `fft` and `spatial` is probably about right given the compile times involved.

Done.

It's also been helpfully pointed out to me that `_core` probably isn't the best name for the extension module. It's very generic. So I renamed it from `_core` to `_fmm_core` to be more clear and searchable.

alugowski commented:
@rgommers glad the wheel sizes are not impacted more than expected. If the size becomes an issue in the future I'm happy to revisit.

alugowski commented:
A small update:

I convinced the threadpoolctl folks to add a way to specify custom controllers. As of threadpoolctl 3.2.0, that is now possible. Now regular FMM can be controlled by threadpoolctl, and so I'm also updating this PR to do the same.

SciPy users already use threadpoolctl to control BLAS and OpenMP parallelism, so it's natural to use the same mechanism here. Since this code isn't public yet, I went ahead and dropped the now-redundant `parallelism` arguments to `mmread` and `mmwrite`.

j-bowhay commented:
Looks like you have picked up an unwanted submodule change for `unuran`.

alugowski commented:
> Looks like you have picked up an unwanted submodule change for `unuran`.

Thank you! Reverted.

alugowski commented:
@rgommers I know it's a big PR, let me know if I can help make it more digestible in any way.

rgommers commented:
Argh no, it's perfectly digestible. I just dropped the ball on final review - too much to do in too little time.

rgommers commented:
> I convinced the threadpoolctl folks to add a way to specify custom controllers. As of threadpoolctl 3.2.0, that is now possible. Now regular FMM can be controlled by threadpoolctl, and so I'm also updating this PR to do the same.
> 
> SciPy users already use threadpoolctl to control BLAS and OpenMP parallelism, so it's natural to use the same mechanism here. Since this code isn't public yet, I went ahead and dropped the now-redundant `parallelism` arguments to `mmread` and `mmwrite`.

This is quite nice! I played with it a bit, and I'm happy with how this looks.

rgommers reviewed (approved):
This looks quite good, CI is happy, performance is great, build time and wheel file size increase seem acceptable and we touched on that on the mailing list before. It's a lot of C++ code and the C++17 usage may perhaps cause an issue on some niche platform - but more review won't catch that, and that this has previously been released as a standalone package should have caught most of the problems here. So let's give this a go!

Very nice work, thanks again @alugowski.

alugowski commented:
Great! Thank you for the invaluable feedback throughout the process @rgommers !
